server/mdm/apple/gdmf/api.go

```diff
@@ -0,0 +1,186 @@
+func GetLatestOSVersion(device apple_mdm.MachineInfo) (*Asset, error) {
+	r, err := GetAssetMetadata()
+	if err != nil {
+		return nil, fmt.Errorf("retrieving asset metadata: %w", err)
+	}
+
+	assetSet := r.PublicAssetSets.MacOS // default to public asset set; note that if the device is not macOS, iPhone, or iPad, we'll fail to match the supported device and return an error below
+	if strings.HasPrefix(device.Product, "iPhone") ||
+		strings.HasPrefix(device.Product, "iPad") ||
+		strings.HasPrefix(device.SoftwareUpdateDeviceID, "iPhone") ||
+		strings.HasPrefix(device.SoftwareUpdateDeviceID, "iPad") {
+		assetSet = r.PublicAssetSets.IOS
+	}
+	latestIdx := -1
+	for i, s := range assetSet {
+		for _, d := range s.SupportedDevices {
+			if d == device.Product || d == device.SoftwareUpdateDeviceID {
+				if latestIdx == -1 {
+					latestIdx = i // first match found, update the index
+					continue
+				}
+				if apple_mdm.CompareVersions(assetSet[latestIdx].ProductVersion, s.ProductVersion) < 0 {
+					latestIdx = i // found a later version, update the index
+				}
+			}
+		}
+	}
+	if latestIdx == -1 {
+		return nil, fmt.Errorf("no matching asset found for device %s", device.Product)
+	}
+	return &assetSet[latestIdx], nil
+}
+
+// client is a package-level client (similar to http.DefaultClient) so it can
+// be reused instead of created as needed, as the internal Transport typically
+// has internal state (cached connections, etc) and it's safe for concurrent
+// use.
+var client = fleethttp.NewClient(fleethttp.WithTimeout(10 * time.Second))
+
+// GetAssetMetadata retrieves the asset metadata from the Apple Software Lookup Service[1][2].
+// [1]: http://gdmf.apple.com/v2/pmv
+// [2]: https://support.apple.com/guide/deployment/use-mdm-to-deploy-software-updates-depafd2fad80/web
+func GetAssetMetadata() (*APIResponse, error) {
+	baseURL := getBaseURL()
+	reqURL, err := url.Parse(baseURL)
+	if err != nil {
+		return nil, fmt.Errorf("parsing base URL: %w", err)
+	}
+
+	req, err := http.NewRequest(http.MethodGet, reqURL.String(), nil)
+	if err != nil {
+		return nil, fmt.Errorf("creating request to Apple endpoint: %w", err)
+	}
+	req.Header.Set("User-Agent", "fleet-device-management")
+
+	resp, err := doWithRetry(req)
+	if err != nil {
+		return nil, fmt.Errorf("retrieving asset metadata: %w", err)
+	}
+	defer resp.Body.Close()
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("reading response body from Apple endpoint: %w", err)
+	}
+	var dest APIResponse
+	if err := json.Unmarshal(body, &dest); err != nil {
+		return nil, fmt.Errorf("decoding response data from Apple endpoint: %w", err)
+	}
+
+	return &dest, nil
+}
+
+func doWithRetry(req *http.Request) (*http.Response, error) {
+	const (
+		maxRetries           = 3
+		retryBackoff         = 1 * time.Second
+		maxWaitForRetryAfter = 10 * time.Second
+	)
+	var resp *http.Response
+	var err error
+	op := func() error {
+		resp, err = client.Do(req)
```

Comment:
This might be a bit tricky to handle as the response is currently not consumed until the caller of `doWithRetry`, but this will leak non-consumed and non-closed response bodies (eg. if the response is not a 200). The response body must be consumed to EOF and closed in order for the client to reuse a TCP connection (https://pkg.go.dev/net/http#Client.Do).

Reply:
maybe in a `defer` (after the `if err != nil` block) that checks if `resp.StatusCode >= 400` then it reads the response and closes it?

Reply:
Thanks! I've updated with the approach you suggested. Do you have any ideas of good ways to test for leaks here?

Reply:
Nice! I don't think it's worth the trouble of testing this specifically - it would be complex (the checks would need to be at the TCP level I believe, something like checking that the same network connection is used across calls), and as long as we don't remove this defer, we can assume that Go will behave as documented.